### ISSUE TITLE
problem combining memory-mapped arrays on the same file

### ISSUE THREAD

**Issue by henrypinkard** [dask/dask#4920]
Hi, I've run into a bug with combining multiple memory-mapped files into a dask array. I wasn't able to figure out if this is a problem with dask or numpy, so posting on both. 

The context is, I have a large binary file that contains several chunks of data that I'd like to combine into a single multidimensional array, without having to immediately read all of these chunks into memory. My strategy for doing so is is to use `np.memmap` to create multiple memory-mapped arrays which are views different parts of the same binary file. Then, I use `da.stack` to make these part of the same dask array. However, when I then try to access the data in the dask array, it copies the second memory-mapped array twice instead of doing the first and then the second. Using dask 1.2.2 and numpy 1.14.5. Reproducing code below:

```
import dask.array as da
import numpy as np

#create a file with 10 random bytes
file = open('demo_data', 'wb')
file.write(b'ashekwicht')
file.close()

#open the file and read the bytes the bytes we just wrote
file = open('demo_data', 'rb')
first10 = np.frombuffer(file.read(10), np.uint8)
print('The first 5 bytes of the file are: {}'.format(first10[:5]))
print('The next 5 bytes of the file are: {}'.format(first10[5:]))

#now create two memory mapped views into the same file
mmap1 = np.memmap(file, dtype=np.uint8, mode='r', offset=0, shape=5)
mmap2 = np.memmap(file, dtype=np.uint8, mode='r', offset=5, shape=5)

#combine them with dask and read the 10 bytes again
dask_array = da.stack((mmap1, mmap2))
print('First 5 byte values {}'.format(np.array(dask_array[0, :])))
print('Next 5 byte values {}'.format(np.array(dask_array[0, :])))
```

**Comment by jcrist** [comment]
Thanks for filing the issue, this is indeed a bug in dask. When dask methods get non-dask collection objects we hash then using a extensible hash function (more on this here: http://docs.dask.org/en/latest/custom-collections.html#implementing-deterministic-hashing). The hash function is called `tokenize`, and is used to determine identical data/tasks to avoid doing extra work.

```python
In [3]: from dask.base import tokenize

In [4]: tokenize(mmap1) == tokenize(mmap2)  # This shouldn't be true
Out[4]: True
```

It looks like the implementation for `np.memmap` objects is failing to properly get the offset, and defaulting to 0 for both, resulting in the same hash for both arrays (and thus the reused result). The faulty code is here:

https://github.com/dask/dask/blob/0554dfbc541fc12c07b0d7611c67c71f5c6b47f4/dask/base.py#L747-L751

On my version of numpy there is an `offset` attribute that we could be using instead. I haven't looked to see when that attribute was added, but if it's in `1.13`, then we may be able to replace those lines with the attribute as is.

Would you be interested in investigating further, and submitting a PR?

**Comment by henrypinkard** [comment]
Yeah happy to take a look

**Comment by henrypinkard** [comment]
Think this should fix it. `offset` is indeed an attribute in numpy `1.13.0`, so just get that when available

**Comment by jcrist** [comment]
Fixed by #4931.

**Comment by henrypinkard** [comment]
@jcrist It seems that this bug still exists in dask and was not fixed by #4931. Although the test added in that PR passes, the example I posted at the top of this thread still doesn't work. Any thoughts?
